### ISSUE TITLE
Invalid output (missing hole) when performing 2d minkowski sum

### ISSUE THREAD

**Issue by olivier-roussel** [CGAL/cgal#2517]
## Issue Details

In some cases, 2d Minkowski Sum produces invalid output.
When performing the Minkowski sum on a polygon with one hole with another (simple) polygon, it seems that some weird conditions on the outer boundary of the polygon with hole lead to invalid output, where the output should still contain an "eroded" hole and this hole is actually missing. See attached pictures that illustrates the case.

![cgal_bug_input](https://user-images.githubusercontent.com/2340432/31350182-834e0cc4-ad26-11e7-9bc0-d27eb93d2920.png)
Input polygons: the polygon with hole (in blue) and the other polygon, the "pattern" (in green).

![cgal_bug_output](https://user-images.githubusercontent.com/2340432/31350185-8634d80a-ad26-11e7-971f-d47a0811ddfd.png)
Ouput of the minkowski sum where a 1x1 hole is missing.

Note that if we reduce the outer boundary of the polygon with hole on which is applied the sum, by replacing each -3 coordinate by -2, we obtain the correct result (the "eroded" hole is present in the output sum).

## Source Code
```
#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
#include <CGAL/basic.h>
#include <CGAL/minkowski_sum_2.h>

  typedef CGAL::Exact_predicates_exact_constructions_kernel     Kernel;
  typedef Kernel::Point_2                                       Point_2;
  typedef CGAL::Polygon_2<Kernel>                               Polygon_2;
  typedef CGAL::Polygon_with_holes_2<Kernel>                    Polygon_with_holes_2;

  Polygon_with_holes_2 polygon_w_holes;
  polygon_w_holes.outer_boundary().push_back(Point_2(-3., -3.));
  polygon_w_holes.outer_boundary().push_back(Point_2(7., -3.));
  polygon_w_holes.outer_boundary().push_back(Point_2(7., 7.));
  polygon_w_holes.outer_boundary().push_back(Point_2(-3., 7.));

  // the hole for the polygon to apply pattern
  Polygon_2 hole;
  hole.push_back(Point_2(5., 2.));
  hole.push_back(Point_2(2., 2.));
  hole.push_back(Point_2(2., 5.));
  hole.push_back(Point_2(5., 5.));
  polygon_w_holes.add_hole(hole);

  Polygon_2 pattern;
  pattern.push_back(Point_2(1., 1.));
  pattern.push_back(Point_2(-1., 1.));
  pattern.push_back(Point_2(-1., -1.));
  pattern.push_back(Point_2(1., -1.));

  Polygon_with_holes_2  sum = CGAL::minkowski_sum_2(polygon_w_holes, pattern);

  std::cout << "outer boundary: " << sum.outer_boundary() << std::endl;
  for(auto it_hole = sum.holes_begin(); it_hole != sum.holes_end(); ++it_hole)
  {
    std::cout << "hole : " << *it_hole << std::endl;
  }
```

Produces output:
```
outer boundary: 8 -5 -5 -3 -5 7 -5 7 -3 7 7 5 7 -5 7 -5 5
 ```

But expected output should look like this:
```
outer boundary: 8 -5 -5 -3 -5 7 -5 7 -3 7 7 5 7 -5 7 -5 5
 hole : 3 3 3 4 4 4 4 3
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Fedora 25 64 bits
* Compiler: g++ 6.3.1
* Release or debug mode: Release
* Specific flags used (if any): -
* CGAL version: 4.9.1 (git hash: 01ce16fd5f6c6f660d39e037faefbb6b3fdabc6c )
* Boost version: 1.60
* Other libraries versions if used (Eigen, TBB, etc.): -

**Comment by efifogel** [comment]
I've reproduced it; seems like a bug.
The output should, actually, be:

outer boundary: 4 -5 7 -5 -5 7 -5 7 7
hole : 4 3 3 3 2 2 2 2 3

You can compute it using the alternative method based on vertical
decomposition:

#include <CGAL/Polygon_vertical_decomposition_2.h>

  CGAL::Polygon_vertical_decomposition_2<Kernel>::Traits_2 traits;
  CGAL::Polygon_vertical_decomposition_2<Kernel> vertical_decomp(traits);
  sum = CGAL::minkowski_sum_2(P, Q, vertical_decomp, traits);

**Comment by olivier-roussel** [comment]
You are right, my code sample was not consistent w.r.t. the pictures. The second polygon (the "pattern") was centered at (-1, -1), and the output you suggested actually correspond to this case. 
I've edited the code accordingly to the expected case, where the square pattern is centered at (0, 0).

The workaround using the vertical decomposition method works great, thanks a lot for the tip.

**Comment by efifogel** [comment]
Hi Sebastian,

Oliver has encountered a bug in the Minkowski sum code based on the reduced
convolution.
I've traced it to:

bool AABB_collision_detector_2::check_collision(const Point_2 &t)

I believe that there are actually 2 bugs there, and I think I know how to
fix. Could you please verify my changes and perhaps suggest improvements.

1. Consider the statements in lines 82-83:

Point_2 t_q = *m_q.outer_boundary().vertices_begin() + Vector_2(ORIGIN, t);
Point_2 t_p = *m_p.outer_boundary().vertices_begin() - Vector_2(ORIGIN, t);

m_q is the original polygon and m_p is the inverted one, so the operation
should be flipped; that is, shift m_q by -t and shift m_p by t.

2. Consider the statement:
    return
      bounded_side_2(m_p.outer_boundary().vertices_begin(),
                     m_p.outer_boundary().vertices_end(), t_q,
                     m_p.outer_boundary().traits_member()) ==
ON_BOUNDED_SIDE ||
      bounded_side_2(m_q.outer_boundary().vertices_begin(),
                     m_q.outer_boundary().vertices_end(), t_p,
                     m_q.outer_boundary().traits_member()) ==
ON_BOUNDED_SIDE;

We check only against the outer boundaries, but one polygon can be
completely inside a hole of the other, so we also need to add tests against
every hole of each polygon; something like the following (haven't tested it
thoroughly):

    bool in_mp =
      bounded_side_2(m_p.outer_boundary().vertices_begin(),
                     m_p.outer_boundary().vertices_end(), t_q,
                     m_p.outer_boundary().traits_member()) ==
ON_BOUNDED_SIDE;
    if (m_p.number_of_holes() == 0) {
      if (in_mp) return true;
    }
    bool in_mq =
      bounded_side_2(m_q.outer_boundary().vertices_begin(),
                     m_q.outer_boundary().vertices_end(), t_p,
                     m_q.outer_boundary().traits_member()) ==
ON_BOUNDED_SIDE;
    if (m_q.number_of_holes() == 0) {
      if (in_mq) return true;
    }
    if (!in_mq && !in_mp) return false;
    if (in_mp) {
      for (auto it = m_p.holes_begin(); it != m_p.holes_end(); ++it) {
        if (bounded_side_2(it->vertices_begin(), it->vertices_end(), t_q,
                           it->traits_member()) == ON_BOUNDED_SIDE)
          return false;
      }
      return true;
    }
    for (auto it = m_q.holes_begin(); it != m_q.holes_end(); ++it) {
      if (bounded_side_2(it->vertices_begin(), it->vertices_end(), t_p,
                         it->traits_member()) == ON_BOUNDED_SIDE)
        return false;
    }
    return true;

Thanks,
Efi

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by afabri** [comment]
Efi has written "Hi Sebastien" but not @sloriot

**Comment by efifogel** [comment]
Yes, I should have responded earlier.
I meant Sebastian (with an 'a') Morr, who is the main author of this code.
I would like S. Morr to respond and stay connected, but this is probably
not the case, and that's ok.
So, there is no need to wait for him.
@Laurent, you can go ahead and process the pull request.

   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
This discussion is an issue. Maybe you are talking about the PR #2571. The PR #2571 is waiting for a reaction of @blinry.

**Comment by lrineau** [comment]
> The PR #2571 is waiting for a reaction of @blinry.

... how is actually Sebastian Morr (strange Github nickname!)

**Comment by efifogel** [comment]
S. Morr was a MsC student of Michael Hemmer in Braunschweig.
S. Morr based his work on the code written by Alon Barnea from TAU.
Before M. Hemmer was in Braunschweig he was at TAU, where he supervised A.
Barnea.
I guess all this is irrelevant. Nor S. Mor neither A. Barnea are involved.
They stopped being involved the minute they graduated.
I asked for S. Morr's opinion mainly to inform him rather than to get
feedback.
The fix is not complicated and local.


   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by lrineau** [comment]
Fixed by #2571.
